File: backend/src/schemas/task.py
"""Task schemas."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class TaskBase(BaseModel):
    """Base task schema."""

    title: str = Field(min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=2000)
    is_completed: bool = False
    priority: str = "medium"


class TaskCreate(TaskBase):
    """Schema for creating a task."""

    tags: list[str] = Field(default=[], description="List of tag names")


class TaskUpdate(BaseModel):
    """Schema for updating a task."""

    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=2000)
    is_completed: Optional[bool] = None
    priority: Optional[str] = None
    tags: Optional[list[str]] = Field(default=None, description="List of tag names")


class TaskResponse(TaskBase):
    """Complete task schema for responses."""

    id: int
    user_id: str
    created_at: datetime
    updated_at: datetime
    tags: list[str] = Field(default=[], description="List of tag names")

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/task.ts
/**
 * Task type definitions
 */

export type Priority = 'high' | 'medium' | 'low';

export interface Task {
  id: number;
  user_id: number;
  title: string;
  description: string | null;
  is_completed: boolean;
  priority: Priority;
  created_at: string;
  updated_at: string;
  tags: string[];
}

export interface TaskCreate {
  title: string;
  description?: string;
  priority?: Priority;
  tags?: string[];
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  priority?: Priority;
  is_completed?: boolean;
  tags?: string[];
}

export interface TaskFilters {
  status?: 'all' | 'completed' | 'pending';
  priority?: Priority;
  tags?: string[];
  search?: string;
  sort_by?: 'created_at' | 'priority' | 'title';
  sort_order?: 'asc' | 'desc';
}
